### PR TITLE
Keep zero check be compatible with different sympy versions

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130729

# Motivation
I found a difference between sympy 1.12 and 1.13.
```python
# for 1.12
>>> import sympy
>>> a = sympy.Number(0.0)
>>> a == 0
True
```
```python
# for 1.13
>>> import sympy
>>> a = sympy.Number(0.0)
>>> a == 0
False
```
The different behavior will impact the result of [safe_mul](https://github.com/pytorch/pytorch/blob/6beec34b1c6803d5f6648c3cd7c262d6432374c8/torch/utils/_sympy/value_ranges.py#L521-L528), resulting in an incorrect results when `a = sympy.Number(0.0)`, `b = inf` and the result is `nan` if sympy version is 1.13. (the expected result is **0**)
```python
def safe_mul(a, b):
    # Make unknown() * wrap(0.0) == wrap(0.0)
    if a == 0.0:
        return a
    elif b == 0.0:
        return b
    else:
        return a * b
```

In different sympy versions, `sympy.Number(0)` always has the same behavior that equals to 0.0.
```python
>>> import sympy
>>> a = sympy.Number(0)
>>> a == 0.0
True # for different sympy versions
```
So, use 0.0 when checking zero in safe_mul to keep compatible with different sympy versions.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Uf, this is surely going to bite us in other places...

**Comment by Skylion007** [comment]
> Uf, this is surely going to bite us in other places...

I guess we should enforce the latest sympy version only?

**Comment by guangyey** [comment]
> > Uf, this is surely going to bite us in other places...
> 
> I guess we should enforce the latest sympy version only?

#130689 intends to enforce to upgrade sympy version to 1.13 or later.

**Review by EikanWang — APPROVED** [review]

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -519,10 +519,10 @@ def mul(cls, a, b):
             return cls.and_(a, b)
 
         def safe_mul(a, b):
-            # Make unknown() * wrap(0) == wrap(0)
-            if a == 0:
+            # Make unknown() * wrap(0.0) == wrap(0.0)
+            if a == 0.0:
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:523]: if a is 0 (integer) does this still match it?

Comment (guangyey) [reply torch/utils/_sympy/value_ranges.py:523]: Yes
```python
>>> import sympy
>>> a = sympy.Number(0)
>>> a == 0.0
True # for different sympy versions
```

Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:523]: 🤔

Comment (guangyey) [reply torch/utils/_sympy/value_ranges.py:523]: It seems that `is_zero` is a better choice.
```python
>>> import sympy
>>> a = sympy.Number(0)
>>> a.is_zero
True
>>> b = sympy.Number(0.0)
>>> b.is_zero
True
```

**Pull request closed without merging**